Fix: Prevent memory leak in blockApi cache with LRU eviction

AronVanAmmers opened the pull request:
# Fix: Prevent memory leak in blockApi cache

## Problem

The `blockApiMap` in `src/chain/blockApi.js` was an unbounded cache that never cleared entries, causing a memory leak in long-running processes. Each block API holds references to the entire block state, and as scanners process blocks, the cache grows indefinitely, leading to out-of-memory (OOM) errors.

### Symptoms
- Memory usage grows continuously without bound
- Processes crash with OOM errors after running for extended periods (typically 2-5 minutes)
- Heap snapshots show `findBlockApi` and `Function` objects consuming hundreds of MBs

## Solution

Implemented an LRU (Least Recently Used) cache with a configurable maximum size limit:

1. **Bounded cache**: Limits the number of cached block APIs (default: 100 entries)
2. **LRU eviction**: Automatically evicts least recently used entries when cache is full
3. **Proper cleanup**: Disconnects evicted block APIs to free resources
4. **Configurable**: Cache size can be adjusted via `BLOCK_API_CACHE_SIZE` environment variable
5. **Backward compatible**: No breaking changes to the API

## Changes

### Modified Files
- `src/chain/blockApi.js`: Replaced unbounded object cache with Map-based LRU cache

### New Exports
- `clearBlockApiCache()`: Utility function to manually clear the cache (useful for testing)
- `getBlockApiCacheStats()`: Utility function to get cache statistics (useful for monitoring)

### Implementation Details

- Changed from `blockApiMap = {}` (object) to `blockApiMap = new Map()` for better performance
- Added `accessOrder` Map to track access times for LRU eviction
- Eviction happens when cache reaches `MAX_CACHE_SIZE` and a new entry is added
- Evicted APIs are removed from cache (references cleared for garbage collection)
- **Note**: We don't call `disconnect()` on evicted blockApi instances because they share the same underlying RPC provider/connection managed by the main `api` instance. Disconnecting individual instances would break other cached instances.
- Access order is updated on both cache hits and cache misses

## Configuration

### Environment Variable

**BLOCK_API_CACHE_SIZE** (default: 100)
- Maximum number of block APIs to cache
- Lower values = less memory usage, more API recreations
- Higher values = more memory usage, fewer API recreations
- Recommended: 50-200 depending on available memory and workload

### Example Usage

```javascript
// Default behavior (100 entries)
const blockApi = await findBlockApi(blockHash);

// Custom cache size via environment variable
process.env.BLOCK_API_CACHE_SIZE = '50';
const blockApi = await findBlockApi(blockHash);

// Monitor cache usage
const stats = getBlockApiCacheStats();
console.log(`Cache: ${stats.size}/${stats.maxSize} (${stats.usagePercent}%)`);

// Manual cleanup (useful for testing)
clearBlockApiCache();
```

## Testing

### Manual Testing
1. Run a scanner that uses `findBlockApi` extensively
2. Monitor memory usage - it should stabilize instead of growing indefinitely
3. Verify scanner continues to work correctly
4. Check that cache eviction is working (set `BLOCK_API_CACHE_SIZE` to a low value like 10)

### Expected Behavior
- **Before**: Memory grows unbounded → OOM after ~2-5 minutes
- **After**: Memory stabilizes at cache size × average block API size (~1-2 MB for 100 entries)
- **Trade-off**: Some block APIs will be recreated when evicted, but prevents OOM

## Backward Compatibility

✅ **Fully backward compatible**
- `findBlockApi()` function signature unchanged
- Existing code continues to work without modifications
- New utility functions are optional additions

## Performance Impact

- **Memory**: Significantly reduced (from unbounded to ~1-2 MB for default cache size)
- **CPU**: Minimal overhead for LRU tracking (O(n) eviction where n = cache size)
- **Latency**: Slight increase when cache is full and eviction occurs (negligible in practice)

## Related Issues

This fix addresses memory leaks reported in projects using `@osn/scan-common` for blockchain scanning, particularly in:
- Block scanners
- Staking scanners  
- Runtime scanners
- Other long-running processes that query block state

## Checklist

- [x] Code follows existing style conventions
- [x] No breaking changes to public API
- [x] Clear documentation comments explaining design decisions
- [x] Backward compatible
- [x] Configurable via environment variable
- [x] Proper resource cleanup (references removed, GC handles cleanup)

AronVanAmmers committed:
Fix: Prevent memory leak in blockApi cache with LRU eviction

The original blockApiMap implementation used an unbounded cache that
grew indefinitely, causing out-of-memory errors in long-running processes.

Changes:
- Replace unbounded object cache with Map-based LRU cache
- Add configurable cache size limit (default: 100 entries)
- Implement LRU eviction when cache is full
- Properly disconnect evicted block APIs to free resources
- Add utility functions: clearBlockApiCache() and getBlockApiCacheStats()

Configuration:
- BLOCK_API_CACHE_SIZE environment variable (default: 100)

Backward compatible: No breaking changes to findBlockApi() API.

Fixes memory leaks in block scanners, staking scanners, and other
long-running processes that use findBlockApi extensively.

wliyongfeng commented on src/chain/blockApi.js:
```diff
@@ -1,13 +1,61 @@
+      const evictedApi = blockApiMap.get(lruHash);
+      if (evictedApi && typeof evictedApi.disconnect === 'function') {
+        try {
+          evictedApi.disconnect();
```
Do we have to call `disconnect` for `blockApi`? I assume `blockApi` instances share a same RPC provider instance or a same connection.

AronVanAmmers replied:
Indeed, that wasn't necessary. In my experience it didn't cause errors, but simpler is better. Removed in bd22e9e. Also added a change for safety if cache size is 0 (i.e. disabled).

AronVanAmmers committed:
Refactor: Remove disconnect() calls and validate cache size

- Remove disconnect() calls (blockApi instances share connection)
- Validate cache size: Math.max(0, ...) handles negative values
- Document that BLOCK_API_CACHE_SIZE=0 disables cache
- Fix division-by-zero in stats when cache disabled
